Add Kayobe Workflows [GitHub] (#753)
stackhpc/stackhpc-kayobe-config: doc/source/configuration/kayobe-automation.rst

=================
Kayobe Automation
=================

What is Kayobe Automation
=========================

> `Kayobe automation <https://github.com/stackhpc/kayobe-automation/>`__ is a collection of scripts and tools that enable the automation of kayobe related operations through the use of CI/CD platforms such as those provided by GitHub and GitLab.

Reviewer: For a first sentence, this is pretty intense. maybe something like:
```suggestion
`Kayobe automation <https://github.com/stackhpc/kayobe-automation/>`__ is a collection of scripts and tools that automate kayobe operations.
It is deployed and controlled by CI/CD platforms such as GitHub actions and GitLab pipelines.
```
My version of the content might not be 100% correct, I'm just suggesting the formatting style more than anything.

> Kayobe automation provides users an easy process of performing tasks such as; overcloud service deploy, config-diff, tempest testing and many more.

Reviewer: ```suggestion
Kayobe automation provides users with an easy process to perform tasks such as; overcloud service deploy, config-diff, tempest testing and many more.
```

With it being integrated into platforms such as GitHub or GitLab it builds a close relationship between the contents of the deployments kayobe configuration and what is currently deployed.
This is because operations such as opening a pull request will trigger a config diff to be generated providing insight on what impact it might have on services or a tempest test that could be scheduled to run daily providing knowledge of faults earlier than before.

> Kayobe automation has been designed to be independent of any CI/CD platform with all tasks running inside of a purpose build Kayobe container.

Reviewer: ```suggestion
Kayobe automation has been designed to be independent of any CI/CD platform with all tasks running inside of a purpose built Kayobe container.
```

However, platform specific workflows need to be deployed to bridge the gap between the contents of Kayobe Automation and these CI/CD platforms.
To achieve this work has been carried to template workflows as deployment specific choices have to be made when writing workflows, such as; multiple environment support, container registry location and secret sharing.

> The templating of workflows is offered through the `stackhpc.kayobe_workflows <https://github.com/stackhpc/ansible-collection-kayobe-workflows/>`__ collection which currently supports GitHub workflows and should enable the easy and error free deployment of workflows.

Reviewer: ```suggestion
The templating of workflows is offered through the `stackhpc.kayobe_workflows <https://github.com/stackhpc/ansible-collection-kayobe-workflows/>`__ collection which currently supports GitHub workflows.
```

GitHub Deployment
=================

To enable Kayobe Automation where GitHub Actions is used please follow the steps described below starting with the deployment the runners.

> Runner Deployment

Reviewer 2: Overall there's a bit too much left to interpretation here around paths. I'd suggest using relative paths within kayobe-config.

Author: Used relative paths.

-----------------

1. Identify a suitable host for hosting the runners.
    GitHub runners need to be deployed on a host which has not had Docker deployed using kolla.
    This is because GitHub runners cannot provide `network options when running in a container <https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontaineroptions>`__.

    Ideally an Infra VM could be used here or failing that the control host.
    Wherever it is deployed the host will need access to the :code:`admin_network`, :code:`public_network` and the :code:`pulp registry` on the seed.

> 2. Edit the environments :code:`inventory/groups` to add the predefined :code:`github-runners` group to :code:`infra-vms`

Reviewer: ```suggestion
2. Edit the environment's :code:`inventory/groups` to add the predefined :code:`github-runners` group to :code:`infra-vms`
```

.. code-block:: ini

    [infra-vms:children]
    github-runners

> 3. Edit the environments :code:`inventory/hosts` to define the host(s) that will host the runners.

Reviewer: ```suggestion
3. Edit the environment's :code:`inventory/hosts` to define the host(s) that will host the runners.
```

.. code-block:: ini

>     [github-runner]

Reviewer 2: ```suggestion
    [github-runners]
```

    runner-01

> 4. Provide all the relevant Kayobe :code:`group_vars` for :code:`github-runners` under :code:`inventory/hosts/github-runners`

Reviewer 2: ```suggestion
4. Provide all the relevant Kayobe :code:`group_vars` for :code:`github-runners` under :code:`inventory/group_vars/github-runners`
```

>     * `infra-vms` ensuring all required `infra_vm_extra_network_interface` are defined

Reviewer 2: ```suggestion
    * `infra-vms` ensuring all required `infra_vm_extra_network_interfaces` are defined
```

    * `network-interfaces`

>     * `python-interpreter.yml` ensuring that `ansible_python_interpreter: /usr/bin/python3` has been set

Reviewer: Wouldn't this be image-dependent?

Author: I don't believe so. The reason for this to be defined would be because the default python interpreter might not be present or the user `runner_user` lacks permission to use.

Reviewer 2: Does the normal method of using the Kayobe virtualenv not work?

      - openstack

    github_runners:
      runner_01: {}
      runner_02: {}
      runner_03: {}

    docker_users:
      - "{{ runner_user }}"

    pip_install_packages:
      - name: docker

If using multiple environments add an extra label to :code:`default_runner_labels` to distinguish these runners from runners belonging to other environments.
Also feel free to change the number of runners and their names.

6. Obtain a personal access token that would enable the registration of GitHub runners against the `github_account` and `github_repo` defined above.
    This token ideally should be `fine-grain personal access token <https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token>`__ which may require the organisation to enable such tokens beforehand steps can be found `here <https://docs.github.com/en/organizations/managing-programmatic-access-to-your-organization/setting-a-personal-access-token-policy-for-your-organization>`__
    The repository permissions for a fine-grain personal access token should be; :code:`Actions: R/W, Administration: R/W, Metadata: R`

>     Once obtained the key add to :code:`secrets.yml` under :code:`secrets_github_access_token`

Reviewer: ```suggestion
    Once the key has been obtained, add it to :code:`secrets.yml` under :code:`secrets_github_access_token`
```

7. Run :code:`kayobe playbook run ${KAYOBE_CONFIG_PATH}/ansible/deploy-github-runner.yml`

> 8. Check runners have registered properly by visiting the repositories :code:`Action` tab -> :code:`Runners` -> :code:`Self-hosted runners`

Reviewer: ```suggestion
8. Check runners have registered properly by visiting the repository's :code:`Action` tab -> :code:`Runners` -> :code:`Self-hosted runners`
```

9. Repeat the above steps for each environment you intend to deploy runners within.
    You can share the fine-grain access token between environments.

Workflow Deployment
-------------------

1. Edit `inventory/group_vars/github-writer/writer.yml` in the base configuration making the appropriate changes to your deployments specific needs see documentation for `stackhpc.kayobe_workflows.github <https://github.com/stackhpc/ansible-collection-kayobe-workflows/tree/main/roles/github>`__

2. Run :code:`kayobe playbook run ${KAYOBE_CONFIG_PATH}/ansible/write-workflows.yml`

> 3. Add all required secrets to repository either via the GitHub UI or GitHub CLI (may require repository owner)

Reviewer: A description of what these are would help here

Author: Description added.

    * KAYOBE_AUTOMATION_SSH_PRIVATE_KEY
    * KAYOBE_VAULT_PASSWORD
    * REGISTRY_PASSWORD
    * TEMPEST_OPENRC

> Note if using multiple environments and are not sharing secrets between environments then each of must have the environment name prefix for each enviromment for example

Reviewer: ```suggestion
Note if you are using multiple environments and not sharing secrets between environments then each of must have the environment name prefix for each enviromment for example
```

    * PRODUCTION_KAYOBE_AUTOMATION_SSH_PRIVATE_KEY
    * PRODUCTION_KAYOBE_VAULT_PASSWORD
    * PRODUCTION_REGISTRY_PASSWORD
    * PRODUCTION_TEMPEST_OPENRC
    * STAGING_KAYOBE_AUTOMATION_SSH_PRIVATE_KEY
    * STAGING_KAYOBE_VAULT_PASSWORD
    * STAGING_REGISTRY_PASSWORD
    * STAGING_TEMPEST_OPENRC

4. Commit and push all newly generated workflows found under :code:`.github/workflows`

Final Steps
-----------

> Some final steps include the following; running config-diff will require that :code:`.automation.conf/config.sh` contains a list :code:`KAYOBE_CONFIG_VAULTED_FILES_PATHS_EXTRA` of all vaulted files contained within the config.

Reviewer: ```suggestion
Some final steps include the following: running config-diff will require that :code:`.automation.conf/config.sh` contains a list :code:`KAYOBE_CONFIG_VAULTED_FILES_PATHS_EXTRA` of all vaulted files contained within the config.
```

All such files can be found with :code:`grep -r "$ANSIBLE_VAULT;1.1;AES256" .` though make sure NOT to include `kolla/passwords.yml` and `secrets.yml`
Also make sure tempest has been configured appropriately in :code:`.automation.conf/config.sh` to meet the limitations of a given deployment such as not using a too high of :code:`TEMPEST_CONCURRENCY` value and that overrides and load/skips lists are correct.

> Finally, once all the workflows and configuration has been pushed and reviewed you can build a kayobe image using `Build Kayobe Docker Image` once succesfully built and pushed to a container registry other workflows can be used.

Reviewer: ```suggestion
Finally, once all the workflows and configuration has been pushed and reviewed you can build a kayobe image using the `Build Kayobe Docker Image` workflow. Once it is successfully built and pushed to a container registry, other workflows can be used.
```

=================
Kayobe Automation
=================

> 

Reviewer 2: I would start with a very high level list of the components of the CI/CD system for context:

* Platform-specific CI/CD workflow definitions
* Platform-specific CD/CD runners
* Platform-independent Kayobe automation tooling

Reviewer 2: Then you could list the supported platforms.

Author: Not sure if what I have added is what you are after.

> What is Kayobe Automation

Reviewer 2: Could you rename this section "Concepts", with a subheading of "Kayobe Automation"

=========================

`Kayobe automation <https://github.com/stackhpc/kayobe-automation/>`__ is a collection of scripts and tools that automate kayobe operations.
It is deployed and controlled by CI/CD platforms such as GitHub actions and GitLab pipelines.

> Kayobe automation provides users with an easy process to perform tasks such as; overcloud service deploy, config-diff, tempest testing and many more.

Reviewer: Minor nit I missed last time
```suggestion
Kayobe automation provides users with an easy process to perform tasks such as: overcloud service deploy, config-diff, tempest testing, and many more.
```

With it being integrated into platforms such as GitHub or GitLab it builds a close relationship between the contents of the deployments kayobe configuration and what is currently deployed.
This is because operations such as opening a pull request will trigger a config diff to be generated providing insight on what impact it might have on services or a tempest test that could be scheduled to run daily providing knowledge of faults earlier than before.

Kayobe automation has been designed to be independent of any CI/CD platform with all tasks running inside of a purpose built Kayobe container.

> However, platform specific workflows need to be deployed to bridge the gap between the contents of Kayobe Automation and these CI/CD platforms.

Reviewer 2: I'd add a new "Workflows" subsection here.

Author: Added subsection

> To achieve this work has been carried to template workflows as deployment specific choices have to be made when writing workflows, such as; multiple environment support, container registry location and secret sharing.

Reviewer 2: ```suggestion
Workflows are templated for each Kayobe configuration repository, ensuring appropriate workflow input parameters are defined, and any necessary customisations can be applied.
```

> The templating of workflows is offered through the `stackhpc.kayobe_workflows <https://github.com/stackhpc/ansible-collection-kayobe-workflows/>`__ collection which currently supports GitHub workflows.

Reviewer 2: Could you add a new "Runners" section here, with a platform-independent intro to runners.

> GitHub Deployment

Reviewer 2: ```suggestion
GitHub Actions
```

=================

> To enable Kayobe Automation where GitHub Actions is used please follow the steps described below starting with the deployment the runners.

Reviewer 2: ```suggestion
To enable CI/CD where GitHub Actions is used please follow the steps described below starting with the deployment of the runners.
```

    Ideally an Infra VM could be used here or failing that the control host.
    Wherever it is deployed the host will need access to the :code:`admin_network`, :code:`public_network` and the :code:`pulp registry` on the seed.

2. Edit the environment's :code:`inventory/groups` to add the predefined :code:`github-runners` group to :code:`infra-vms`

.. code-block:: ini

    [infra-vms:children]
    github-runners

3. Edit the environment's :code:`inventory/hosts` to define the host(s) that will host the runners.

.. code-block:: ini

    [github-runner]
    runner-01

4. Provide all the relevant Kayobe :code:`group_vars` for :code:`github-runners` under :code:`inventory/hosts/github-runners`
    * `infra-vms` ensuring all required `infra_vm_extra_network_interface` are defined
    * `network-interfaces`
    * `python-interpreter.yml` ensuring that `ansible_python_interpreter: /usr/bin/python3` has been set

5. Create `runner.yml` file which will contain the variables required to deploy a series of runners

> .. code-block:: yaml

Reviewer 2: Should we just provide this file in the repo with some sensible defaults and placeholders for required variables?

Author: I have added the file to `${KAYOBE_CONFIG_PATH}/inventory/group_vars/github-runners/runners.yml`

    ---
    runner_user: VM_USER_NAME_HERE
    github_account: ORG_NAME_HERE
    github_repo: KAYOBE_CONFIG_REPO_NAME_HERE
    access_token: "{{ secrets_github_access_token }}"

    base_runner_dir: /opt/actions-runner

    default_runner_labels:
      - kayobe
      - openstack

>     github_runners:

Reviewer 2: The format of this variable should be documented.

Author: Added format of variable to documentation.

      runner_01: {}
      runner_02: {}
      runner_03: {}

    docker_users:
      - "{{ runner_user }}"

    pip_install_packages:
      - name: docker

> If using multiple environments add an extra label to :code:`default_runner_labels` to distinguish these runners from runners belonging to other environments.

Reviewer 2: Let's be clear that the label needs to be `{{ kayobe_environment }}`, or better include it in the defaults.

Author: Would something like this be suitable.

```yaml
default_runner_labels:
  - kayobe
  - openstack
  - {{ kayobe_environment | default(omit) }}
```

Reviewer 2: You'll need quotes around the braces. I don't know if `omit` will do the right thing - it evaluates to a strange `omit_glsjlfszlknglksnaklgns` string that may be used as a label. You could try just using an empty string, if the role will use that correctly.

Author: I ran the playbook with the `default(omit)` and it was correctly ignored if running from the base config

Also feel free to change the number of runners and their names.

6. Obtain a personal access token that would enable the registration of GitHub runners against the `github_account` and `github_repo` defined above.

>     This token ideally should be `fine-grain personal access token <https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token>`__ which may require the organisation to enable such tokens beforehand steps can be found `here <https://docs.github.com/en/organizations/managing-programmatic-access-to-your-organization/setting-a-personal-access-token-policy-for-your-organization>`__

Reviewer 2: ```suggestion
    This token ideally should be `fine-grained personal access token <https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token>`__ which may require the organisation to enable such tokens beforehand.
    Steps can be found `here <https://docs.github.com/en/organizations/managing-programmatic-access-to-your-organization/setting-a-personal-access-token-policy-for-your-organization>`__.
```

>     The repository permissions for a fine-grain personal access token should be; :code:`Actions: R/W, Administration: R/W, Metadata: R`

Reviewer 2: ```suggestion
    The repository permissions for a fine-grained personal access token should be; :code:`Actions: R/W, Administration: R/W, Metadata: R`
```

    Once the key has been obtained, add it to :code:`secrets.yml` under :code:`secrets_github_access_token`

> 7. Run :code:`kayobe playbook run ${KAYOBE_CONFIG_PATH}/ansible/deploy-github-runner.yml`

Reviewer 2: If the VM is actually an infra VM, we're missing some steps. You could at least link to the upstream infra VM docs.

Author: Link has been added to upstream documentation.

8. Check runners have registered properly by visiting the repository's :code:`Action` tab -> :code:`Runners` -> :code:`Self-hosted runners`

9. Repeat the above steps for each environment you intend to deploy runners within.

>     You can share the fine-grain access token between environments.

Reviewer 2: ```suggestion
    You can share the fine-grained access token between environments.
```

Workflow Deployment
-------------------

> 1. Edit `inventory/group_vars/github-writer/writer.yml` in the base configuration making the appropriate changes to your deployments specific needs see documentation for `stackhpc.kayobe_workflows.github <https://github.com/stackhpc/ansible-collection-kayobe-workflows/tree/main/roles/github>`__

Reviewer 2: ```suggestion
1. Edit `inventory/group_vars/github-writer/writer.yml` in the base configuration making the appropriate changes to your deployments specific needs. See documentation for `stackhpc.kayobe_workflows.github <https://github.com/stackhpc/ansible-collection-kayobe-workflows/tree/main/roles/github>`__.
```

2. Run :code:`kayobe playbook run ${KAYOBE_CONFIG_PATH}/ansible/write-workflows.yml`

3. Add all required secrets to repository either via the GitHub UI or GitHub CLI (may require repository owner)
    * KAYOBE_AUTOMATION_SSH_PRIVATE_KEY: private key used by Ansible to authenticate with machines.
    * KAYOBE_VAULT_PASSWORD: password used by the config to encrypt Ansible Vault secrets.
    * REGISTRY_PASSWORD: password used to login to the docker registry such as Pulp.
    * TEMPEST_OPENRC: contents of :code:`kolla/public-openrc.sh`

> Note if you are using multiple environments and not sharing secrets between environments then each of must have the environment name prefix for each enviromment for example

Reviewer 2: ```suggestion
Note if you are using multiple environments and not sharing secrets between environments then each of these must have the environment name prefix for each environment, for example:
```

    * PRODUCTION_KAYOBE_AUTOMATION_SSH_PRIVATE_KEY
    * PRODUCTION_KAYOBE_VAULT_PASSWORD
    * PRODUCTION_REGISTRY_PASSWORD
    * PRODUCTION_TEMPEST_OPENRC
    * STAGING_KAYOBE_AUTOMATION_SSH_PRIVATE_KEY
    * STAGING_KAYOBE_VAULT_PASSWORD
    * STAGING_REGISTRY_PASSWORD
    * STAGING_TEMPEST_OPENRC

4. Commit and push all newly generated workflows found under :code:`.github/workflows`

Final Steps
-----------

Some final steps include the following: running config-diff will require that :code:`.automation.conf/config.sh` contains a list :code:`KAYOBE_CONFIG_VAULTED_FILES_PATHS_EXTRA` of all vaulted files contained within the config.
All such files can be found with :code:`grep -r "$ANSIBLE_VAULT;1.1;AES256" .` though make sure NOT to include `kolla/passwords.yml` and `secrets.yml`
Also make sure tempest has been configured appropriately in :code:`.automation.conf/config.sh` to meet the limitations of a given deployment such as not using a too high of :code:`TEMPEST_CONCURRENCY` value and that overrides and load/skips lists are correct.

> Finally, once all the workflows and configuration has been pushed and reviewed you can build a kayobe image using the `Build Kayobe Docker Image` workflow. Once it is successfully built and pushed to a container registry, other workflows can be used.

Reviewer 2: This could use more info on when to rebuild the Kayobe image.

Author: Added some information on why you might want to rebuild the image.
